# Exercise notebook 1: Having a go at it

This Jupyter notebook, for the first part of The Open University's _Learn to Code for Data Analysis_ course, contains code examples and coding activities for you.

You'll come across steps in the course directing you to this notebook. Once you've done each exercise, go back to the corresponding step and mark it as complete. 

In [49]:
# this code conceals irrelevant warning messages
import warnings
warnings.simplefilter('ignore', FutureWarning)

## Exercise 1: variables and assignments

A **variable** is a named storage for values. An **assignment** takes a value (like the number 100 below) and stores it in a variable (`deathsInPortugal` below).

In [50]:
deathsInPortugal = 100

To display the value stored in a variable, write the name of the variable. 

In [51]:
deathsInPortugal

100

Each variable can store one value at any time, but the value stored can vary over time, by assigning a new value to the variable.

In [52]:
deathsInPortugal = 100
deathsInPortugal = 140
deathsInPortugal

140

Each assignment is written on a separate line. The computer executes the assignments one line at a time, from top to bottom.

In [53]:
deathsInPortugal = 140
deathsInAngola = 6900


In [54]:
deathsInBrazil = 4400
deathsInRussia = 17000
deathsInIndia = 240000
deathsInChina = 41000
deathsInSA = 25000

### Task

Add assignments to the code cell above (or in a new code cell) for the estimated deaths by TB in 2013 in the remaining BRICS countries. The values are as follows: Russia 17000, India 240000, China 41000, South Africa 25000. 

Don't forget to run the code cell, so that the new variables are available for the exercises further below.

**Now go back to the course step and mark it complete.**

## Exercise 2: expressions

An **expression** is a fragment of code that has a value. A variable name, by itself, is an expression: the expression's value is the value stored in the variable. **In Jupyter notebooks, if the last line of a code cell is an expression, then the computer will show its value when executing the cell.** 

In [36]:
deathsInPortugal

140

By contrast, a **statement** is a command for the computer to do something. Commands don't produce values, and therefore the computer doesn't display anything.

In [55]:
deathsInPortugal = 140

More complex expressions can be written using the arithmetic **operators** of addition (`+`), substraction (`-`), multiplication (`*`) and division (`/`). For example, the total number of deaths in the three countries is:

In [38]:
deathsInAngola + deathsInBrazil + deathsInPortugal

11440

If the calculated value needs to be used later on in the code, it has to be stored in a variable. In general, the right-hand side of an assignment is an expression; its value is calculated (the expression is **evaluated**) and stored in the variable.

In [39]:
totalDeaths = deathsInAngola + deathsInBrazil + deathsInPortugal
totalDeaths

11440

The average number of deaths is the total divided by the number of countries.

In [40]:
totalDeaths / 3

3813.3333333333335

The average could also be calculated with a single expression.

In [41]:
(deathsInAngola + deathsInBrazil + deathsInPortugal) / 3

3813.3333333333335

The parentheses (round brackets) are necessary to state that the sum has to be calculated before the division. Without parentheses, Python follows the conventional order used in mathematics: divisions and multiplications are done before additions and subtractions.

In [42]:
deathsInAngola + deathsInBrazil + deathsInPortugal / 3

11346.666666666666

### Task

- In the cell below, write code to calculate the total number of deaths in the five BRICS countries (Brazil, Russia, India, China, South Africa) in 2013. Run the code to see the result, which should be 327400.

In [43]:
deathsInBRICS = deathsInBrazil + deathsInRussia + deathsInIndia +deathsInChina + deathsInSA
deathsInBRICS

327400

- In the cell below, write code to calculate the average number of deaths in the BRICS countries in 2013. Run the code to see the result, which should be 65480.

In [44]:
avgDeaths = deathsInBRICS/5
avgDeaths

65480.0

Cindy Lorenzi - comment
Was just looking at the results for the average of the deaths in the BRICS countries and the result shows up as a float unless I actually transform it into an int. This is rather expected as division in Python does that - but are we expected to transform the output so that it becomes an integer? (i.e. something like this

In [45]:
modDeaths = int(deathsInBRICS/5)
modDeaths

65480

**Now go back to the course step and mark it complete.**

1.10 Functions

Functions
After the total and the average, next on my to-do list is to calculate the largest number of deaths. This will be the maximum. It takes another single line of code to calculate it.

In [46]:
max(deathsInAngola, deathsInBrazil, deathsInPortugal)

6900

In [47]:
In this expression, max() is a function – the parenthesis are a reminder that the name max doesn’t refer to a variable.
A function is a piece of code that calculates (returns) a value, given zero or more values (the function’s arguments). 

SyntaxError: invalid syntax (<ipython-input-47-1649ff15428f>, line 1)

## Exercise 3: functions quiz 

A **function** takes zero or more values (the function's **arguments**) and **returns** (produces) a value. To **call** (use) a function, write the function name, followed by its arguments within parentheses (round brackets). Multiple arguments are separated by commas. Function names follow the same rules and conventions as variable names. A function call is an expression: the expression's value is the value returned by the function.

Python provides two functions to compute the **maximum** (largest) and **minimum** (smallest) of two or more values.

In [ ]:
max(deathsInBrazil, deathsInPortugal)

In [ ]:
min(deathsInAngola, deathsInBrazil, deathsInPortugal)

The **range** of a set of values is the difference between the maximum and the minimum.

In [ ]:
largest = max(deathsInBrazil, deathsInRussia, deathsInIndia, deathsInChina, deathsInSA)
smallest = min(deathsInBrazil, deathsInRussia, deathsInIndia, deathsInChina, deathsInSA)
deathsRange = largest - smallest
deathsRange

### Tasks

Answer the quiz questions in the course. All of them can be answered by editing the above code cell. Don't forget that you can use TAB-completion to quickly write the variable names of the remaining BRICS countries, namely Russia, India, China and South Africa (Brazil is already in the code above).

**Quiz**
1) If the function name is misspelled as Min, what kind of error is it?
A) Name Error # The computer will understand than Min(…, …) is a function call but doesn’t know of any function with that name. Remember that names are case-sensitive.

2) If a parenthesis or comma is forgotten, what kind of error is it?
A) Syntax Error # A function call requires two parentheses around the arguments, and one comma between successive arguments. Forgetting any of them therefore deviates from the syntax of the Python language.

In [ ]:
1.12 Comments

A comment starts with a hash sign (#) and goes until the end of the line. Computers ignore all comments, they just execute the code. Comments are your insurance policy: they help you understand your own code if you come back to it after a long break.

## Exercise 4: comments

**Comments** start with the hash sign (`#`) and go until the end of the line. They're used to annotate the code, e.g. to indicate the units of values.

In [ ]:
# population unit: thousands of inhabitants
populationOfPortugal = 10608

# deaths unit: inhabitants
deathsInPortugal = 140

# deaths per 100 thousand inhabitants
deathsInPortugal * 100 / populationOfPortugal

### Task

Calculate the deaths per 100 thousand inhabitants for Brazil. Its population in 2013 was roughly 200 million and 362 thousand people. You should obtain a result of around 2.2 deaths per 100 thousand people.

**Now go back to the course step and mark it complete.**

In [ ]:
# population unit: thousands of inhabitants
populationOfBrazil = 200362

# deaths unit: inhabitants
deathsInBrazil = 4400

# deaths per 100 thousand inhabitants
deathsInBrazil * 100 / populationOfBrazil

**Values have units**
Before I move on, let me explain the importance of using comments to record units of measurement.

Values are not just numbers, they have units: degrees Celsius, number of inhabitants, thousands of gallons, etc. Always make a note of the units the value refers to, using comments. This makes it easier to check whether the expressions are right. Disregarding the units will lead to wrong calculations and results.

## Exercise 5: pandas quiz

All programs in this course must start with the following **import statement**, to load all the code from the pandas **module**.

In [56]:
from pandas import *

The words in boldface (`from` and `import`) are **reserved words** of the Python language; they cannot be used as names.

### Task

Answer the quiz questions in the course. You can change the above line of code to find out the answers.

1)What kind of error will you get if you misspell ‘pandas’ as ‘Pandas’?

A) The computer is expecting a name but there is no module with the name ‘Pandas’. Remember that names are case-sensitive.

2)What kind of error will you get if you misspell ‘import’ as ‘impart’?

A) The computer is expecting a reserved word and anything else will raise a syntax error.

3) What kind of error will you get if you forget the asterisk?

A) The statement cannot end with the reserved word ‘import’; the computer is expecting an indication of what to import.

**1.16 This week’s data**
For the next part of the course you’ll need to use a data file.

I have created a table with all the data necessary for the project and saved it in an Excel file. Excel is a popular application to create, edit and analyse tabular data. You won’t need Excel to complete this course, but many datasets are provided as Excel files.

The data file is ‘WHO POP TB some.xls’, which is in the folder you have previously dowloaded (and uploaded to CoCalc, if the case). The file is encoded using UTF-8, a character encoding that allows for accented letters. **Do not open or edit the file, as you may change how it is encoded**, which will lead to errors later on. If you do want to look at its contents, make a copy of the file and look at the copy.

Let’s proceed to loading the data – you’ll learn how to do this in the next step.

**1.17 Loading the data**
Loading the data
Many applications can read files in Excel format, and pandas can too. Asking the computer to read the data looks like this:


In [57]:
data = read_excel('WHO POP TB some.xls')
data

Country  Population (1000s)  TB deaths
0                  Angola               21472       6900
1                  Brazil              200362       4400
2                   China             1393337      41000
3       Equatorial Guinea                 757         67
4           Guinea-Bissau                1704       1200
5                   India             1252140     240000
6              Mozambique               25834      18000
7                Portugal               10608        140
8      Russian Federation              142834      17000
9   Sao Tome and Principe                 193         18
10           South Africa               52776      25000
11            Timor-Leste                1133        990

The variable name data is not descriptive, but as there is only one dataset in our analysis, there is no possible confusion with other data, and short names help to keep the lines of code short.

The function read_excel() takes a file name as an argument and returns the table contained in the file. In pandas, tables are called **dataframes**. To load the data, I simply call the function and store the returned dataframe in a variable.

A file name must be given as a string, a piece of text surrounded by quotes. The quote marks tell Python that this isn’t a variable, function or module name. Also, the quote marks state that this is a single name, even if it contains spaces, punctuation and other characters besides letters.

Misspelling the file name, or not having the file in the same folder as the notebook containing the code, results in a file not found error. In the example below, an ‘o’ in the filename was replaced by an ‘a’, leading to an error error message saying that no such file can be found.

**Jupyter notebooks show strings in red. If you see red characters until the end of the line, you have forgotten to type the second quote that marks the end of the string.**

@SurajPandeyAs long as you keep the notebook and the data file in the same location you do not need to give an absolute/complete path.

The relative path in python is always the same location as where the code file is located.

So, if both your notebook ('Exercises_1.ipynb') and the data file (WHO POP TB some.xls') are in the same location ('E:\py example\open university'), then you can read the file with the following code: 
data = read_excel('WHO POP TB some.xls') 
instead of having to provide the full path.

If you had put the notebook in 'E:\py example\open university' and the data file in 'E:\py example\open university\data', then to read the data file you would need the following code: 
data = read_excel('WHO POP TB some.xls\data')

## Exercise 6: selecting a column

The `read_excel()` function takes a **string** with the name of an Excel file, and returns a **dataframe**, the pandas representation of a table. The computer reports a **file not found** error if the file is not in the same folder as this notebook, or the file name is misspelt.

In [58]:
data = read_excel('WHO POP TB some.xls')
data

Country  Population (1000s)  TB deaths
0                  Angola               21472       6900
1                  Brazil              200362       4400
2                   China             1393337      41000
3       Equatorial Guinea                 757         67
4           Guinea-Bissau                1704       1200
5                   India             1252140     240000
6              Mozambique               25834      18000
7                Portugal               10608        140
8      Russian Federation              142834      17000
9   Sao Tome and Principe                 193         18
10           South Africa               52776      25000
11            Timor-Leste                1133        990

The expression `dataFrame[columnName]` evaluates to the column with the given name (a string). Column names are case sensitive. Misspelling the column name will result in a rather long **key error** message. You can see what happens by changing the string in the next code cell (e.g. replace `TB` by `tb`) and running it. Don't forget to undo your change and run the code again.

**1.18** Strings are verbatim text, which means that the column name must be written exactly as given in the dataframe, which you saw after loading the data. The slightest deviation leads to a key error, which can be seen as a kind of name error. You can try out in the Week 1 exercise notebook what happens when misspelling the column name. **The error message is horribly long. In such cases, just skip to the last line of the error message to see the type of error.**

In [59]:
tbColumn = data['TB deaths']
tbColumn

0       6900
1       4400
2      41000
3         67
4       1200
5     240000
6      18000
7        140
8      17000
9         18
10     25000
11       990
Name: TB deaths, dtype: int64

### Task

In the next cell, select the population column and store it in a variable (you'll use it in the next exercise). You need to scroll back to the start of the exercise to see the column's name.

In [60]:
popColumn = data['Population (1000s)']
popColumn

0       21472
1      200362
2     1393337
3         757
4        1704
5     1252140
6       25834
7       10608
8      142834
9         193
10      52776
11       1133
Name: Population (1000s), dtype: int64

**Now go back to the course step and mark it complete.**

Calculations on a column
The max() or sum() functions are general functions that can operate on lists of values such as the contents of a single dataframe column. More efficient calculations can be performed on pandas dataframes using methods. A method is a function that can only be called in a certain context. In this course, the context will mostly be a dataframe or a column. A method call looks like a function call, but adds the context in which to call the method: context.methodName(argument1, argument2, ...). In other words, a dataframe method can only be called on dataframes, a column method only on columns. **Because methods are functions, a method call returns a value and is therefore an expression.**

## Exercise 7: calculations on a column


A **method** is a function that can only be called in a certain context, like a dataframe or a column. A **method call** is of the form `context.methodName(argument1, argument2, ...)`.   

Pandas provides several column methods, including to calculate the sum, the largest, and the smallest of the numbers in a column, as follows.

In [61]:
tbColumn.sum()

354715

In [62]:
tbColumn.max()

240000

In [63]:
tbColumn.min()

18

The **mean** of a collection of numbers is the sum of those numbers divided by how many there are.

In [64]:
tbColumn.sum() / 12

29559.583333333332

In [65]:
tbColumn.mean()

29559.583333333332

The **median** of a collection of numbers is the number in the middle, i.e. half of the numbers are below the median and half  are above.

In [66]:
tbColumn.median()

5650.0

The mean is five times higher than the median. While half the countries had less than 5650 deaths in 2013, some countries had far more, which pushes the mean up.

The median is probably closer to the intuition you have of what ‘average’ should mean (pun intended). News reports don’t always make clear what average measure is being used, and using the mean may distort reality. For example, the mean household income in a country will be influenced by very poor and very rich households, whereas the median income doesn’t take into account how poor or rich the extremes are: it will always be half the households below and half above the median.

### Tasks

Use the population column variable from the previous exercise to calculate:

- the total population

In [67]:
popColumn.sum()

3103150

- the maximum population

In [68]:
popColumn.max()

1393337

- the minimum population

In [69]:
popColumn.min()

193

In [70]:
popRange = (popColumn.max() - popColumn.min())*1000
popRange

1393144000

**Now go back to the course step and mark it complete.**

Like sum(), the column methods min() and max() don’t need arguments, whereas the Python functions min() and max() did need them, because there was no context (column) providing the values.

## Exercise 8: sorting on a column

The dataframe method `sort_values()` takes as argument a column name and returns a new dataframe, with rows in ascending order according to the values in that column.

In [ ]:
data.sort_values('TB deaths')

**Sorting doesn't change the original table.**

In [ ]:
data # rows still in original order

Sorting on a column that has text will put the rows in alphabetical order.

In [ ]:
data.sort_values('Country')

### Task

Sort the same table by population, to quickly see which are the least and the most populous countries.

In [ ]:
data.sort_values('Population (1000s)')

**Now go back to the course step and mark it complete.**

1.21 The last remaining task is to calculate the death rate of each country.
You may recall that with the simple approach I’d have to write:

rateAngola = deathsInAngola * 100 / populationOfAngola
rateBrazil = deathsInBrazil * 100 / populationOfBrazil
and so on, and so on. If you’ve used spreadsheets, it’s the same process: create the formula for the first row and then copy it down for all the rows. This is laborious and error-prone, e.g. if rows are added later on. Given that data is organised by columns, wouldn’t it be nice to simply write the following?

rateColumn = deathsColumn * 100 / populationColumn
Say no more: your wish is pandas’s command.

## Final quiz: Calculations over columns

This information will help you to answer questions in the final quiz.

The value of an arithmetic expression involving columns is a column. In evaluating the expression, the computer computes the expression for each row.

In [72]:
deathsColumn = data['TB deaths']
populationColumn = data['Population (1000s)']
rateColumn = deathsColumn * 100 / populationColumn
rateColumn

0     32.134873
1      2.196025
2      2.942576
3      8.850727
4     70.422535
5     19.167186
6     69.675621
7      1.319759
8     11.901928
9      9.326425
10    47.370017
11    87.378641
dtype: float64

Tadaaa! With pandas, the arithmetic operators become much smarter. When adding, subtracting, multiplying or dividing columns, the computer understands that the operation is to be done row by row and creates a new column.

To add a new column to a dataframe, 'select' a non-existing column, i.e. with a new name, and assign to it.

In [73]:
data['TB deaths (per 100000)'] = rateColumn
data

Country  Population (1000s)  TB deaths  \
0                  Angola               21472       6900   
1                  Brazil              200362       4400   
2                   China             1393337      41000   
3       Equatorial Guinea                 757         67   
4           Guinea-Bissau                1704       1200   
5                   India             1252140     240000   
6              Mozambique               25834      18000   
7                Portugal               10608        140   
8      Russian Federation              142834      17000   
9   Sao Tome and Principe                 193         18   
10           South Africa               52776      25000   
11            Timor-Leste                1133        990   

    TB deaths (per 100000)  
0                32.134873  
1                 2.196025  
2                 2.942576  
3                 8.850727  
4                70.422535  
5                19.167186  
6                69.675621  
7                 1.319759  
8                11.901928  
9                 9.326425  
10               47.370017  
11               87.378641

### Tasks

Add code to calculate:

- the range of the population, in thousands of inhabitants

In [74]:
rangePop = popColumn.max() - popColumn.min()
rangePop

1393144

- the mean of the death rate

In [75]:
rateColumn.mean()

30.223859369031903

- the median of the death rate

In [76]:
rateColumn.median()

15.534556945151177

Now you can answer the questions in the **final quiz**.

Writing up the analysis
Once you’ve done your analysis, you may want to record it or share it with others. The best way is to write up what you’ve discovered.

There is no right or wrong way to write up data analysis but the important thing is to present the answers to the questions you had. To keep things simple, I suggest the following structure:

1.A descriptive title
2.An introduction setting the context and stating what you want to find out with the data.
3.A section detailing the source(s) of the data, with the code to load it into the notebook.
4.One or more sections showing the processes (calculating statistics, sorting the data, etc.) necessary to address the questions.
5.A conclusion summarising your findings, with qualitative analysis of the quantitative results and critical reflection on any shortcomings in the data or analysis process.
6.You don’t need to explain your code, but it’s helpful to write the text in such a way that even readers who know nothing about Python or pandas can follow your analysis.

You can see how I’ve written up the analysis by opening this week’s project notebook, which is in the folder you downloaded earlier this week.

Summary
This week you learned how to use Jupyter notebooks to write and execute simple programs with Python and the pandas module. You have seen how to load a table from an Excel file, select a column, and compute some simple statistics (like the total, minimum and median) about it. You have also learned how to create a new column with values calculated from other columns. And finally you have seen how to sort a table by one of its columns.

Good programmers don’t just write code, they write readable code, and I’ve insisted on the use of clear variable names and the use of comments. Last but not least, you have received some guidance on how to write up a data analysis and how to share it.

Next week you will learn further ways to manipulate dataframes, in particular to clean data. You will also produce your first data chart, showing variations of values over time.